dor/providers/parsers.py

```diff
@@ -0,0 +1,132 @@
+from datetime import datetime
+import uuid
+
+from utils.element_adapter import ElementAdapter
+from .models import *
+
+
+class DescriptorFileParser:
+    namespaces: dict[str, str] = {
+        "METS": "http://www.loc.gov/METS/v2",
+        "PREMIS": "http://www.loc.gov/premis/v3",
+    }
+
+    def __init__(self, descriptor_path):
+        text = descriptor_path.read_text()
+        self.tree: ElementAdapter = ElementAdapter.from_string(text, self.namespaces)
+
+    def get_id(self):
+        return uuid.UUID(self.tree.get("OBJID"))
+
+    def get_type(self):
+        hdr = self.tree.find("METS:metsHdr")
+        return hdr.get("TYPE")
+
+    def get_alternate_identifier(self):
+        alt_record_id = self.tree.find("METS:metsHdr/METS:altRecordID")
+        if alt_record_id:
+            return AlternateIdentifier(
+                type=alt_record_id.get("TYPE"), id=alt_record_id.text
+            )
+
+    def get_preservation_events(self) -> list[PreservationEvent]:
+        return [self.get_event(elem) for elem in self.tree.findall(".//PREMIS:event")]
+
+    def get_event(self, elem) -> PreservationEvent:
+        event_identifier = elem.find(".//PREMIS:eventIdentifierValue").text
+        event_type = elem.find(".//PREMIS:eventType").text
+        event_datetime = elem.find(".//PREMIS:eventDateTime").text
+        event_detail = elem.find(".//PREMIS:eventDetail").text
+        agent_role = elem.find(".//PREMIS:linkingAgentIdentifierType").text
+        agent_address = elem.find(".//PREMIS:linkingAgentIdentifierValue").text
+        return PreservationEvent(
+            identifier=event_identifier,
+            type=event_type,
+            datetime=datetime.fromisoformat(event_datetime),
+            detail=event_detail,
+            agent=Agent(address=agent_address, role=agent_role),
+        )
+
+    def get_metadata_files(self) -> list[FileMetadata]:
+        return [
+            self.get_md_file_metadatum(elem)
+            for elem in self.tree.findall(".//METS:md[METS:mdRef]")
```

Comment:
Instead of returning a list can we use yield generator, so that it can perform lazy evaluation in case of large data?

Reply:
Possibly. But the objects will typically only have COMMON and DESCRIPTIVE metadata files to start.